### PR TITLE
[ENG-552] Fix Broken Social Links

### PR DESCRIPTION
**Pull request by UdayVarkhedkar**
## Purpose

Fix broken social links (i.e. links to Twitter, Github, etc.) presented as contextual information for users. The impetus for the fix was broken social links when attempting to add a contributor to a project (the links were relative and would result in requests to invalid osf.io pages). 

This problem was later narrowed to be specific to social accounts for which OSF (on the backend) allows for multiple accounts (Twitter, Github, LinkedIn). Broken social links were also appearing for some but not all accounts on the general search page of OSF. 

## Changes

The OSFUser social_links property has been modified to provide a formatted url for all social fields excepting profile websites (previously, the social accounts that allowed for multiple accounts were not providing a formatted url but rather, the account id) which don't need to be formatted. The social_links property was also modified to enforce the types defined in the OSFUser's SOCIAL_FIELDS dictionary. 

The changes have the most direct implications on the OSFUser model and ElasticSearch functions around user information (`search_contributors`, `search`, and `update_user` in website.search.elastic_search.py). 

### Users-Social Links
Users with different types of social links (3 types: profile websites, single-account (i.e. Google Scholar, ORCID, etc.), multiple-account (i.e. Twitter, Github, LinkedIn)) should be tested for the following API endpoints and OSF Pages.

#### API
The following endpoints should be tested to ensure that the appropriate social_links data is being returned. 
* GET `/api/v1/search/user` (uses `search` in elastic_search.py)
* GET `/api/v1/user/search` (uses `search_contributors` in elastic_search.py)

#### Pages
The following pages (there may be other pages with social links I'm not aware of) should be tested to ensure that the social links on the pages are all working properly:
* Add Contributors Modal (off of the contributors page for a project)
* General Search Page
* Profile Page

The following pages should be tested to ensure that social links data is updated appropriately after the form on the page is submitted:
* Profile Social Settings Page

## Documentation

N/A 

## Side Effects

The social_links data returned for the multiple-account services will now return a string (url) rather than a list (of account ids). This may have implications on the way the data is presented on specific pages depending on the javascript on the page. 

Additionally, the way social_link data stored in ElasticSearch will be slightly different. Because the change in data storage is necessary for the general search page to provide valid links, user social links in ElasticSearch will need to be re-indexed (a ticket will be created for this). 

## Ticket

https://openscience.atlassian.net/browse/ENG-552

### CODE REVIEW

**Comment by UdayVarkhedkar** [comment]
> First PR 🎉 🎉 🎉 🎉 !
> 
> Generally this is the type of fix I'd expect see, however there are two minor issues:
> 
> * No tests, please add a test! it's difficult to see what something is trying to fix without a test.
> * Very minor, `basestring` is not Py3 compatible, though I believe in this case just changing to `str` will be sufficient. It will be easy to tell once their's a test.

As far as I can tell, there isn't a good way to check for both unicode and string in python 2 in a way that compatible with python 3. `str` would work in python 3 but in python 2 doesn't include the unicode type. We could use something like the six library and implement a similar solution to the top stack-overflow answer here (https://stackoverflow.com/questions/11301138/how-to-check-if-variable-is-string-with-python-2-and-3-compatibility). What do you think works best/is the most compatible with your work on converting the codebase, John?

**Comment by Johnetordoff** [comment]
I lean on [this document](https://python-future.org/compatible_idioms.html#basestring) for compatibility issues. I like using `past` here.

File: osf/models/user.py [hunk 1]
```diff
@@ -456,10 +457,17 @@ def social_links(self):
...
+                    else:
+                        # Provide a single account url for services which we allow multiple-accounts
+                        social_user_fields[key] = self.SOCIAL_FIELDS[key].format(val[0])
                 else:
-                    social_user_fields[key] = self.SOCIAL_FIELDS[key].format(val)
+                    if not isinstance(val, basestring):
+                        social_user_fields[key] = val
+                    else:
+                        social_user_fields[key] = [val]
         return social_user_fields
```
Comment (pattisdr) [inline osf/models/user.py]: I'm thinking we can improve readability a bit, it takes a minute to break down what is happening.  To start, I'd recommend parallel structure, in the first if, you're checking if it *is* a basestring, in the second conditional, you're checking if it's *not* a basestring. I'd make them the same.

File: osf/models/user.py [hunk 2]
```diff
@@ -456,10 +457,17 @@ def social_links(self):
...
+                        social_user_fields[key] = self.SOCIAL_FIELDS[key].format(val)
+                    else:
+                        # Provide a single account url for services which we allow multiple-accounts
+                        social_user_fields[key] = self.SOCIAL_FIELDS[key].format(val[0])
                 else:
-                    social_user_fields[key] = self.SOCIAL_FIELDS[key].format(val)
+                    if not isinstance(val, basestring):
+                        social_user_fields[key] = val
+                    else:
+                        social_user_fields[key] = [val]
```
Comment (pattisdr) [inline osf/models/user.py]: Is this a possible state, currently?

File: osf/models/user.py [hunk 3]
```diff
@@ -456,10 +457,17 @@ def social_links(self):
         social_user_fields = {}
         for key, val in self.social.items():
             if val and key in self.SOCIAL_FIELDS:
-                if not isinstance(val, basestring):
-                    social_user_fields[key] = val
+                if isinstance(self.SOCIAL_FIELDS[key], basestring):
+                    if isinstance(val, basestring):
+                        social_user_fields[key] = self.SOCIAL_FIELDS[key].format(val)
+                    else:
+                        # Provide a single account url for services which we allow multiple-accounts
```
Comment (pattisdr) [inline osf/models/user.py]: Good place for a comment, but I think small tweaks can make it more clear what you're doing

- "Only return the first URL for services where we allow multiple accounts", or similar.

File: osf_tests/test_user.py [hunk 4]
```diff
@@ -2110,21 +2110,23 @@ def test_profile_website_unchanged(self):
...
+            'github': ['CenterForOpenScience'],
+            'scholar': 'ztt_j28AAAAJ'
         }
         self.user.save()
         assert self.user.social_links == {
             'profileWebsites': ['http://cos.io/'],
             'twitter': 'http://twitter.com/OSFramework',
-            'github': 'http://github.com/CenterForOpenScience'
+            'github': 'http://github.com/CenterForOpenScience',
+            'scholar': 'http://scholar.google.com/citations?user=ztt_j28AAAAJ'
```
Comment (pattisdr) [inline osf_tests/test_user.py]: 🌻

File: osf/models/user.py (hunk 2 above)
Comment (UdayVarkhedkar) [reply osf/models/user.py]: The first one (The value should be an array and is an array in self.social) is with profile websites. The latter (The value should be an array but is an a string in self.social) is not a current existing case.

File: osf/models/user.py (hunk 1 above)
Comment (UdayVarkhedkar) [reply osf/models/user.py]: I attempted to address this with the parallel structure and also added a doc string to clarify the purpose of the function.

**Comment by brianjgeiger** [comment]
When you have a moment, could you check the test and update with develop if need be?

**Pull request merged**